dev/fattest.simplicity/src/componenttest/topology/impl/LibertyServer.java

```diff
@@ -5958,6 +6110,69 @@ public Set<String> getInstalledAppNames(String... possiblyInstalledAppNames) thr
+     * mpFaultTolerance-4.0, servlet-6.0, timedexit-1.0].
+     *
+     * @return           a set of the features installed at runtime
+     * @throws Exception
+     */
+    public Set<String> getInstalledFeatures() throws Exception {
+        Set<String> installedFeatures = new HashSet<>();
+
+        for (String line : findStringsInLogs(INSTALL_FEATURE_MESSAGE_PREFIX)) {
+            installedFeatures.addAll(getInstalledFeaturesFromLogMessage(line));
+            break; //only look at the first message
+        }
+
+        return installedFeatures;
+    }
+
+    /**
+     * Returns a set of the features which were installed at runtime, based on the CWWKF0012I message from the messages.log file.
+     *
+     * Only a line containing a CWWKF0012I message should be passed in. It may be prefixed with timestamps etc.
+     * e.g.
+     * [04/07/2024, 15:26:31:119 BST] 00000035 com.ibm.ws.kernel.feature.internal.FeatureManager A
+     * CWWKF0012I: The server installed the following features: [bells-1.0, cdi-4.0, componenttest-2.0,
+     * concurrent-3.0, jndi-1.0, mpConfig-3.1, mpContextPropagation-1.3, mpFaultTolerance-4.0, servlet-6.0, timedexit-1.0].
+     *
+     * @return           a set of the features installed at runtime
+     * @throws Exception
+     */
+    public static final Set<String> getInstalledFeaturesFromLogMessage(String installedFeaturesMessage) {
+        Set<String> installedFeatures = new HashSet<>();
+        int prefixIndex = installedFeaturesMessage.indexOf(INSTALL_FEATURE_MESSAGE_PREFIX);
+        if (prefixIndex > -1) {
+            String trimmed = installedFeaturesMessage.substring(prefixIndex); //strip off the date & time etc
+            int openBracketIndex = trimmed.indexOf("[");
+            if (openBracketIndex > -1) {
+                trimmed = trimmed.substring(openBracketIndex + 1); //strip off everything up to and including the first "["
+                int closeBracketIndex = trimmed.indexOf("]");
+                if (closeBracketIndex > -1) {
+                    trimmed = trimmed.substring(0, closeBracketIndex);// remove "]."
+                    String[] features = trimmed.split(", ");
+                    for (String feature : features) {
+                        installedFeatures.add(feature);
+                    }
+                } else {
+                    Log.warning(LibertyServer.class, "Installed Features Message CWWKF0012I was not formatted as expected: " + installedFeaturesMessage);
+                }
+            } else {
+                Log.warning(LibertyServer.class, "Installed Features Message CWWKF0012I was not formatted as expected: " + installedFeaturesMessage);
+            }
+        } else {
+            Log.warning(LibertyServer.class, "Installed Features Message CWWKF0012I was not formatted as expected: " + installedFeaturesMessage);
+        }
+        return installedFeatures;
```

Comment:
Just putting out a warning here?

If we end up returning an empty set, that means we do no checking of the features in the server, right?

Reply:
This isn't a test to check the format of that message... I am hoping there are other tests for that. If these warnings do get output then it is probably due to misuse of the method.

Correct, returning an empty set results in no feature checking.